agents_infra/agents_infra/utils/communication.py

```diff
@@ -0,0 +1,183 @@
+        """
+        Initialize AgentCommunication.
+
+        Args:
+            auth_token_type (str): Type of auth token to use in requests.
+            post_data_fn (callable): Function to send data to a controller.
+            controller_urls (list): List of controller URLs in priority order.
+        """
+        self.post_data_fn = post_data_fn
+        self.auth_token_type = auth_token_type
+        self.controller_urls = controller_urls
+        self.current_controller = controller_urls[0]
+
+    @property
+    def logger(self):
+        """
+        Returns a logger named after the agent config or falls back
+        to 'agent-communication'.
+        """
+        name = 'agent-communication'
+        if hasattr(self.post_data_fn.__self__.config, 'name'):
+            name = self.post_data_fn.__self__.config.name
+        return logging.getLogger(name)
+
+    def _switch_controller(self, new_url):
+        """
+        Switch the current controller to a new one.
+        """
+        self.current_controller = new_url
+        self.last_success_time = get_current_time()
+        maybe_log_message(
+            'Controller switched: %s -> %s' % (
+                self.current_controller,
+                new_url
+            ),
+            logger=self.logger,
+            level=logging.INFO
+        )
+
+    def ensure_active_controller(self, api_key):
+        """
+        Ensure there is an active, healthy controller.
+
+        Returns:
+            str or None: Active controller URL, or None if none are healthy.
+        """
+        if self.current_controller != self.controller_urls[0]:
+            self.current_controller = self.try_revert_primary_controller(
+                api_key=api_key
+            )
+
+        if ping_url(
+                self.current_controller,
+                api_key=api_key,
+                logger=self.logger,
+                auth_token_type=self.auth_token_type
+        ):
+            return self.current_controller
+
+        remaining_urls = self._get_lower_priority_urls()
+        healthy_url = find_first_healthy_url(
+            urls=remaining_urls,
+            api_key=api_key,
+            auth_token_type=self.auth_token_type,
+            logger=self.logger
+        )
+        if healthy_url:
+            self._switch_controller(healthy_url)
+            return healthy_url
+
+        maybe_log_message(
+            'No available controller. All health checks failed.',
+            logger=self.logger
+        )
+        return None
+
+    def try_revert_primary_controller(self, api_key):
+        """
+        Attempt to revert back to the primary controller if healthy.
+
+        Returns:
+            str: Current controller URL after the check.
+        """
+        if self.current_controller == self.controller_urls[0]:
+            return self.current_controller
+
+        elapsed = get_current_time() - self.last_success_time
```

Comment:
If `self.last_success_time` might not be set, handle the case to avoid `AttributeError`.